### PR TITLE
Asset Processor - Intermediate Assets - Automated Tests

### PR DESCRIPTION
**Pull request by amzn-mike**
Adding automated tests for intermediate assets feature, plus a couple fixes to issues found during testing.
Also added an automated test for the File Processor to verify it handles duplicate paths in different scanfolders correctly (deletes the correct one instead of all matching relative paths)

### CODE REVIEW

**Review by AMZN-stankowi — APPROVED** [review]

File: Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp [hunk 1]
```diff
@@ -17,91 +17,94 @@ namespace UnitTests
...
+            entry.m_scanFolderPK = m_scanFolder.m_scanFolderID;
+            m_fileEntries.push_back(entry);
         }
     }
 
     void FileProcessorTests::TearDown()
     {
-        m_data->m_databaseLocationListener.BusDisconnect();
-        m_data.reset();
+        m_databaseLocationListener.BusDisconnect();
```
Comment (lmbr-pip) [inline Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp:70]: Theres a lot happening in the Setup call after BusConnect, typically TearDown and Setup are inverses of each other. Double check that things shouldn't be happening prior to BusDisconnect

File: Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp [hunk 2]
```diff
@@ -17,91 +17,94 @@ namespace UnitTests
...
         ConnectionBus::Handler::BusDisconnect(ConnectionBusId);
         AssetProcessorTest::TearDown();
     }
 
     size_t FileProcessorTests::Send([[maybe_unused]] unsigned int serial, [[maybe_unused]] const AzFramework::AssetSystem::BaseAssetProcessorMessage& message)
     {
-        m_data->m_messagesSent++;
+        m_messagesSent++;
```
Comment (lmbr-pip) [inline Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp:77]: nit: prefer prefix increment (not creating extra work for the compiler)

File: Code/Tools/AssetProcessor/native/tests/assetmanager/AssetManagerTestingBase.h [hunk 3]
```diff
@@ -20,10 +20,19 @@
 
 namespace UnitTests
 {
-    class MockDiskSpaceResponder : public AssetProcessor::DiskSpaceInfoBus::Handler
+    struct MockDiskSpaceResponder : AssetProcessor::DiskSpaceInfoBus::Handler
```
Comment (lmbr-pip) [inline Code/Tools/AssetProcessor/native/tests/assetmanager/AssetManagerTestingBase.h:23]: This seems to be a change just to make everything public by default? Unsure of the point of this.

File: Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp [hunk 4]
```diff
@@ -17,91 +17,94 @@ namespace UnitTests
         AssetProcessorTest::SetUp();
         ConnectionBus::Handler::BusConnect(ConnectionBusId);
 
-        m_data.reset(new StaticData());
-        m_data->m_databaseLocationListener.BusConnect();
+        m_databaseLocationListener.BusConnect();
 
-        m_data->m_temporarySourceDir = QDir(m_data->m_temporaryDir.path());
+        m_temporarySourceDir = QDir(m_temporaryDir.path());
```
Comment (lmbr-pip) [inline Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp:22]: As an observation it seems strange that we are doing actual I/O in unit tests. Seems odd that this is not mocked rather than real directory objects with real files.

File: Code/Tools/AssetProcessor/native/tests/assetmanager/IntermediateAssetTests.cpp [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+        ProcessFileMultiStage(3, true);
+
+        AZ::IO::SystemFile::Delete(deleteFilePath);
+        m_assetProcessorManager->AssessDeletedFile(deleteFilePath);
+        RunFile(0); // Process the delete
+
+        // Reprocess the file
+        m_jobDetailsList.clear();
+
+        // AssessModifiedFile is going to set up a OneShotTimer with a 1ms delay on it.  We have to wait a short time for that timer to
```
Comment (lmbr-pip) [inline Code/Tools/AssetProcessor/native/tests/assetmanager/IntermediateAssetTests.cpp:427]: nit: 2ms delay, comment says 1ms. Am curious how we avoid to sleeping in future.

File: Code/Tools/AssetProcessor/native/tests/assetmanager/IntermediateAssetTests.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ */
+
+#pragma once
+
+#include <native/tests/assetmanager/AssetManagerTestingBase.h>
+
+namespace UnitTests
+{
+    struct IntermediateAssetTests
```
Comment (lmbr-pip) [inline Code/Tools/AssetProcessor/native/tests/assetmanager/IntermediateAssetTests.h:15]: Nit: prefer classes to structs so you are thinking about encapsulation, Even in test code it starts making the exposed interface of the object larger and prevents resuse.

**Review by lmbr-pip — COMMENTED** [review]
LG. An awesome amount of test code in here.

Some very minor comments

File: Code/Tools/AssetProcessor/native/tests/assetmanager/IntermediateAssetTests.cpp (hunk 5 above)
Comment (amzn-mike) [reply Code/Tools/AssetProcessor/native/tests/assetmanager/IntermediateAssetTests.cpp:427]: 1ms extra to be sure the timer has elapsed.  I'm not sure we can do much about this, and it shouldn't be as unreliable as the usual sleep people have complained about.  I suppose this could still be flaky if for some reason, sleeping 2ms here does not cause the timer to count as elapsed, but I don't see any reason that would be the case.

The only thing I can think to fix this is to replace the actual code.  I'm not sure why that timer exists in the first place since I'm not too well versed in Qt, I'm guessing its to give the UI some breathing room to update?  I could file a ticket to investigate that at some point if we want?

File: Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp (hunk 2 above)
Comment (amzn-mike) [reply Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp:77]: I'll update it, this is existing code though

File: Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp (hunk 1 above)
Comment (amzn-mike) [reply Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp:70]: The tests are passing currently without any errors showing up, so my vote is to leave it as-is for now.  I don't think we really need to manually clear anything out since its all handled by smart pointers and this test now shuts down the allocators after the destructor runs (which is not how this test was when it was originally written, hence the m_data struct, but we've since moved away from that design)

File: Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp (hunk 4 above)
Comment (amzn-mike) [reply Code/Tools/AssetProcessor/native/tests/FileProcessor/FileProcessorTests.cpp:22]: Yeah, unfortunately there's a lot of that across the unit tests.

File: Code/Tools/AssetProcessor/native/tests/assetmanager/AssetManagerTestingBase.h (hunk 3 above)
Comment (amzn-mike) [reply Code/Tools/AssetProcessor/native/tests/assetmanager/AssetManagerTestingBase.h:23]: Fixed

File: Code/Tools/AssetProcessor/native/tests/assetmanager/IntermediateAssetTests.h (hunk 6 above)
Comment (amzn-mike) [reply Code/Tools/AssetProcessor/native/tests/assetmanager/IntermediateAssetTests.h:15]: Fixed

**Review by lmbr-pip — APPROVED** [review]
